### PR TITLE
Error handling

### LINKED ISSUE #58: Improve error handling
Instead of using error strings, see if we can use something like the Rust error enums.

### PR DESCRIPTION
**Pull request by toni-calvin**
This PR aims to improve error handling and create a structured way to create and throw errors on go code.

### CODE REVIEW

File: pkg/errors/errors.go [hunk 1]
```diff
@@ -0,0 +1,124 @@
+package main
```
Comment (gabrielbosio) [inline pkg/errors/errors.go]: I think we can remove this file now

File: pkg/runners/cairo_runner.go [hunk 2]
```diff
@@ -20,6 +20,8 @@ type CairoRunner struct {
 	mainOffset    uint
 }
 
+var ErrInvalidBuiltin = errors.New("invalid builtin")
```
Comment (gabrielbosio) [inline pkg/runners/cairo_runner.go]: Please, capitalize this error message to keep consistency

File: pkg/vm/vm_core.go [hunk 3]
```diff
@@ -1,18 +1,32 @@
...
+var ErrTraceNotRelocated = errors.New("trace not relocated")
+var ErrUnconstraintRes = errors.New("UnconstrainedRes")
+var ErrDiffAssertValues = errors.New("diff assert values")
+var ErrCantWriteReturnPC = errors.New("can't write return pc")
+var ErrCantWriteReturnFP = errors.New("can't write return fp")
+var ErrFailedToComputeOperandAddress = errors.New("failed to compute operand address")
+var ErrNotRelocatableValue = errors.New("not a relocatable value")
+var ErrNotFeltValue = errors.New("not a felts value")
+var ErrComputeResRelocatableMul = errors.New("compute res relocatable mul")
+var UnconstrainedResAssertEq = errors.New("UnconstrainedResAssertEq")
```
Comment (gabrielbosio) [inline pkg/vm/vm_core.go]: I think it could be better if all error messages are capitalized to keep consistency

File: pkg/vm/vm_test.go [hunk 4]
```diff
@@ -436,9 +436,10 @@ func TestFibonacci(t *testing.T) {
...
 	// TODO: Uncomment test when we have the bare minimum `CairoRun`
 	// err := vm.CairoRun(compiledProgram.Data)
-	// if err != nil {
-	// 	t.Errorf("Program execution failed with error: %s", err)
-	// }
+	//
+	//	if err != nil {
+	//		t.Errorf("Program execution failed with error: %s", err)
+	//	}
 }
```
Comment (gabrielbosio) [inline pkg/vm/vm_test.go]: I think we can remove this test

File: pkg/vm/vm_core.go [hunk 5]
```diff
@@ -1,18 +1,32 @@
...
 
 type VirtualMachineError struct {
 	Msg string
 }
 
+var ErrWrongInstructionEncoding = errors.New("Wrong instruction encoding")
+var ErrFailedToFetchInstruction = errors.New("Failed to fetch instruction in address")
+var ErrNoRelocationFoundSegment = errors.New("No relocation found for execution segment")
+var ErrTraceNotRelocated = errors.New("Trace not relocated")
+var ErrUnconstraintRes = errors.New("UnconstrainedRes")
```
Comment (fmoletta) [inline pkg/vm/vm_core.go:20]:
```suggestion
var ErrUnconstraintRes = errors.New("UnconstrainedRes")
```
This is not descriptive enough

File: pkg/vm/vm_core.go [hunk 6]
```diff
@@ -1,18 +1,32 @@
...
+var ErrDiffAssertValues = errors.New("Diff assert values")
+var ErrCantWriteReturnPC = errors.New("Can't write return pc")
+var ErrCantWriteReturnFP = errors.New("Can't write return fp")
+var ErrFailedToComputeOperandAddress = errors.New("Failed to compute operand address")
+var ErrNotRelocatableValue = errors.New("Not a relocatable value")
+var ErrNotFeltValue = errors.New("Not a felts value")
```
Comment (fmoletta) [inline pkg/vm/vm_core.go]:
```suggestion
var ErrNotFeltValue = errors.New("Not a felt value")
```

File: pkg/vm/vm_core.go [hunk 7]
```diff
@@ -1,18 +1,32 @@
...
+var ErrComputeResRelocatableMul = errors.New("Compute res relocatable mul")
```
Comment (fmoletta) [inline pkg/vm/vm_core.go:27]: This is not descriptive enough

File: pkg/vm/vm_core.go [hunk 8]
```diff
@@ -1,18 +1,32 @@
...
+var UnconstrainedResAssertEq = errors.New("UnconstrainedResAssertEq")
```
Comment (fmoletta) [inline pkg/vm/vm_core.go:28]: This is not descriptive enogh

**Comment by jrchatruc** [comment]
Closing this out as it has become outdated, we will tackle error handling later on in a new PR

**Pull request closed without merging**